### PR TITLE
Item refactoring

### PR DESCRIPTION
**Pull request by ChloeDawn**
The entire package is covered for the most part, bar one or two methods, and cases where an Enigma bug is preventing mapping. Various changes have been made to base methods to better describe their purpose.

**Class name changes:**

- `BaseMapItem`, `isMap` -> `NetworkSyncedItem`, `isNetworkSynced`: See thread https://github.com/FabricMC/yarn/pull/712#discussion_r287575071
  - It is used for items that can send an update packet to the client on each player tick
- `FoodSetting(s)` -> `FoodComponent(s)`: See thread https://github.com/FabricMC/yarn/pull/712#discussion_r287574907
- `BaseBowItem` -> `RangedWeaponItem`
  - The base class shared by bows and crossbows that provides a projectile predication API

**Noteworthy changes in Item/ItemStack:**
- What was previously named "durability" is now "max damage". It represents the maximum amount of damage that can be applied to a stack before it breaks.
  - The inherent problem with describing the maximum amount of damage allowed as "durability" is that durability can easily imply "uses remaining", whereas in this context it describes the total amount of times an item can be damaged before it breaks, regardless of what the stacks current damage is.
- The concept of "amount" is now "count" as this is the Mojang name for this attribute of stacks.
- The method for increasing the stack's count is now `increment`, and for decreasing, `decrement`.
- What was previously "stack size" is now "max count". It represents the maximum count that an item can be stacked up to.

### CODE REVIEW

**Comment by ChloeDawn** [comment]
~~I do not believe `ProjectileShootingItem` accurately represents the class, as it provides no actual shooting logic, only an API for querying held projectiles, and testing ItemStack's for projectile viability. I am open to better name suggestions.~~

**Comment by ChloeDawn** [comment]
I strongly believe that `ItemPropertyGetter` is a wildly incorrect name, but my attempts at discussing it on Discord were fruitless, and responders either were not interested in discussion or believed the class name was fine. The API is intended for computing values that are then tested against expected minimum values declared in item model overrides. It is a client-side API bar the actual class file and the method for registering an implementation that is called from item constructors. The biggest issue I have with the current API naming is that it implies they are general un-sided item properties, but they are not.

However, as there was no luck in my attempts to discuss this, I have left the class name as-is, but have updated the accessors of the API in the Item class to better represent the name.

File: mappings/net/minecraft/item/Item.mapping
Comment (Juuxel) [inline mappings/net/minecraft/item/Item.mapping:112]: Should `Block.matches` and `Fluid.matches` be also changed?

Comment (ChloeDawn) [reply mappings/net/minecraft/item/Item.mapping:112]: Yes, however, it would be out of scope for this pull request; I intend to cover blocks and fluids separately, to avoid a huge mess of changes in one ticket.

File: mappings/net/minecraft/item/CompassItem.mapping
Comment (Gegy) [inline mappings/net/minecraft/item/CompassItem.mapping]: Recall this was brought up on Discord, but not sure where that went. `getAngleToSpawn`?

File: mappings/net/minecraft/block/dispenser/ItemDispenserBehavior.mapping
Comment (Pyrofab) [inline mappings/net/minecraft/block/dispenser/ItemDispenserBehavior.mapping:2]: Is this change just to be in line with `spawnParticles` ? The method does dispense the item.

File: mappings/net/minecraft/item/CompassItem.mapping
Comment (ChloeDawn) [reply mappings/net/minecraft/item/CompassItem.mapping]: Resolved by 54d9bb3

File: mappings/net/minecraft/item/CrossbowItem.mapping
Comment (Gegy) [inline mappings/net/minecraft/item/CrossbowItem.mapping]: Would these `velocityMultiplier` names work as just `speed`?

File: mappings/net/minecraft/item/ArmorStandItem.mapping
Comment (Pyrofab) [inline mappings/net/minecraft/item/ArmorStandItem.mapping]: Why would you change that ? The parameter is not any entity, it's an armor stand.

File: mappings/net/minecraft/item/ArrowItem.mapping
Comment (Pyrofab) [inline mappings/net/minecraft/item/ArrowItem.mapping]: Isn't that less descriptive than previously ?

File: mappings/net/minecraft/item/CrossbowItem.mapping
Comment (ChloeDawn) [reply mappings/net/minecraft/item/CrossbowItem.mapping]: Possibly? It determines a multiplier that is applied to an existing projectile velocity. If you think speed is appropriate for that I will change it. See `ProjectileEntity#setVelocity`.

Comment (ChloeDawn) [reply mappings/net/minecraft/item/CrossbowItem.mapping]:
```java
private static float getVelocityMultiplier(ItemStack stack) {
    if (stack.getItem() == Items.CROSSBOW && hasProjectile(stack, Items.FIREWORK_ROCKET)) {
        return 1.6f;
    }
    return 3.15f;
}
```
```java
new Vec3d(x, y, z).normalize().add(this.random.nextGaussian() * 0.007499999832361937 * trajectoryMultiplier, this.random.nextGaussian() * 0.007499999832361937 * trajectoryMultiplier, this.random.nextGaussian() * 0.007499999832361937 * trajectoryMultiplier).multiply((double)velocityMultiplier);
```

File: mappings/net/minecraft/item/Item.mapping
Comment (Gegy) [inline mappings/net/minecraft/item/Item.mapping:75]: Should this also drop the `on` prefix?

File: mappings/net/minecraft/item/BlockItem.mapping
Comment (Pyrofab) [inline mappings/net/minecraft/item/BlockItem.mapping:18]: Is this normal ?

File: mappings/net/minecraft/item/Item.mapping
Comment (ChloeDawn) [reply mappings/net/minecraft/item/Item.mapping:75]: I retained `on` prefix for event-based methods as it is the current standard in Yarn afaik. Would rather that change be a global migration rather than create inconsistencies now I think.

Comment (ChloeDawn) [reply mappings/net/minecraft/item/Item.mapping:75]: If you are referring to the differing name pattern for `finishUsing`, it is because that is not an event listener, it is a transformer that returns a "finished" stack. See `FoxEntity#tickMovement` when the fox eats food, for example.

File: mappings/net/minecraft/item/CrossbowItem.mapping
Comment (Gegy) [reply mappings/net/minecraft/item/CrossbowItem.mapping]: I think `speed` works nicely here.
Also, for `trajectoryMultiplier`: could maybe be something like `inaccuracy` or `deviation`? Not sure, but the current name is fine too.

File: mappings/net/minecraft/item/Item.mapping
Comment (Gegy) [reply mappings/net/minecraft/item/Item.mapping:75]: Ah, yeah, had thought the prefix was being dropped in other places. 👍

File: mappings/net/minecraft/item/ComplexItem.mapping
Comment (Pyrofab) [inline mappings/net/minecraft/item/ComplexItem.mapping]: Unsure about this name. It's an item that can send a packet every tick when it is in a player's inventory.

File: mappings/net/minecraft/item/CrossbowItem.mapping
Comment (ChloeDawn) [reply mappings/net/minecraft/item/CrossbowItem.mapping]: `inaccuracy` was considered (I think it is also the name MCP opted for), but I don't believe it represents the parameter well. The default value for full precision is `1.0`, and if I saw an "inaccuracy" I would expect `0.0` to be full precision. It is very clearly a multiplier imo.

File: mappings/net/minecraft/item/DefaultFoodSettings.mapping
Comment (Pyrofab) [inline mappings/net/minecraft/item/DefaultFoodSettings.mapping]: Should we name `Items` `DefaultItems` as well ?

File: mappings/net/minecraft/item/ArmorStandItem.mapping
Comment (ChloeDawn) [reply mappings/net/minecraft/item/ArmorStandItem.mapping]: Ah, this was from my initial pass at parameter mappings, cleaning up with generalized names. Do you think `stand` would suffice within this context?

File: mappings/net/minecraft/item/ArrowItem.mapping
Comment (ChloeDawn) [reply mappings/net/minecraft/item/ArrowItem.mapping]: Another case of my initial generalized name pass. I will rename this to `shooter`.

File: mappings/net/minecraft/item/ArmorStandItem.mapping
Comment (Pyrofab) [reply mappings/net/minecraft/item/ArmorStandItem.mapping]: sure

File: mappings/net/minecraft/item/ArrowItem.mapping
Comment (ChloeDawn) [reply mappings/net/minecraft/item/ArrowItem.mapping]: Resolved by 07f735c

File: mappings/net/minecraft/item/DefaultFoodSettings.mapping
Comment (ChloeDawn) [reply mappings/net/minecraft/item/DefaultFoodSettings.mapping]: Good point. Maybe we should re-evaluate the overall class name. Do you think `FoodMetadata` could work? Then again.. the plural `FoodMetadatas` feels a little odd.

File: mappings/net/minecraft/block/dispenser/ItemDispenserBehavior.mapping
Comment (ChloeDawn) [reply mappings/net/minecraft/block/dispenser/ItemDispenserBehavior.mapping:2]: The semantics of the methods in that class are actually rather interesting. I agree that this is likely a poor name choice I've made, but I don't think `dispenseItem` would be a good choice either, as it is too ambiguous alongside `dispense` (deobfuscated name) with identical signature. The difference is this method does not produce a sound or particles. Maybe it should be `dispenseSilently`?

File: mappings/net/minecraft/item/ArmorStandItem.mapping
Comment (ChloeDawn) [reply mappings/net/minecraft/item/ArmorStandItem.mapping]: Resolved by 60cd923

File: mappings/net/minecraft/item/ComplexItem.mapping
Comment (ChloeDawn) [reply mappings/net/minecraft/item/ComplexItem.mapping]: The reasoning for this class name is listed in the initial ticket message. I am open to suggestions for a different name, though. I also agree it is fairly ambiguous. I could not think of an appropriate name to describe the behaviour at the time though, so opted for the Mojang name used in Bedrock.

Data provided by @kashike through Discord:
```
02a08829 T ComplexItem::ComplexItem(std::string const&, int)
02a08829 T ComplexItem::ComplexItem(std::string const&, int)
02a08845 T ComplexItem::~ComplexItem()
02a08841 T ComplexItem::~ComplexItem()
02a08841 T ComplexItem::~ComplexItem()
```

File: mappings/net/minecraft/item/CrossbowItem.mapping
Comment (gegy1000) [reply mappings/net/minecraft/item/CrossbowItem.mapping]: Though, `1.0` wouldn't be full precision based on that usage?

File: mappings/net/minecraft/item/DefaultFoodSettings.mapping
Comment (Pyrofab) [reply mappings/net/minecraft/item/DefaultFoodSettings.mapping]: Good question. `FoodMetadata` seems alright, alternatives are `FoodInformation` and `FoodComponent`.

File: mappings/net/minecraft/item/BlockItem.mapping
Comment (ChloeDawn) [reply mappings/net/minecraft/item/BlockItem.mapping:18]: It is a synthetic method with mapped parameter `value`:

```java
.map(value -> state.with(property, value))
```

File: mappings/net/minecraft/item/HorseArmorItem.mapping
Comment (Pyrofab) [inline mappings/net/minecraft/item/HorseArmorItem.mapping:8]: I feel like this could be more descriptive. Maybe emphasize that it is an "in world" texture ? Or that it is for when it is worn.

File: mappings/net/minecraft/item/CrossbowItem.mapping
Comment (Gegy) [reply mappings/net/minecraft/item/CrossbowItem.mapping]: `1.0` wouldn't be full precision, though, would it? Maybe I'm missing something, I haven't looked into it at all.
(Oops, double posted. Somehow thought GitHub hadn't sent it lol)

File: mappings/net/minecraft/item/Item.mapping
Comment (Pyrofab) [inline mappings/net/minecraft/item/Item.mapping]: I don't think you can say a sword attacks a zombie, except as a figure of speech. Would `hurt` make more sense ?

File: mappings/net/minecraft/item/Item.mapping
Comment (Pyrofab) [inline mappings/net/minecraft/item/Item.mapping]: Maybe this should emphasize that it happens only in a player inventory. A mistake seems too easy.

File: mappings/net/minecraft/item/ItemStack.mapping
Comment (Pyrofab) [inline mappings/net/minecraft/item/ItemStack.mapping:10]: This should be at least `holdingFrame`. `frame` doesn't give any more information than the field type.

File: mappings/net/minecraft/item/ItemStack.mapping
Comment (Pyrofab) [inline mappings/net/minecraft/item/ItemStack.mapping:28]: What was the issue with `flag` ?

File: mappings/net/minecraft/item/CrossbowItem.mapping
Comment (ChloeDawn) [reply mappings/net/minecraft/item/CrossbowItem.mapping]: The following is appended to the velocity for a trajectory. A multiplier of `1.0` means no change:
`
.add(this.random.nextGaussian() * 0.008 * trajectoryMultiplier, this.random.nextGaussian() * 0.008 * trajectoryMultiplier, this.random.nextGaussian() * 0.008 * trajectoryMultiplier)
`

File: mappings/net/minecraft/item/SpawnEggItem.mapping
Comment (Pyrofab) [inline mappings/net/minecraft/item/SpawnEggItem.mapping]: Much needed change. Although it is a static method, so it could be made simpler. Would something like `SpawnEggItem.getAll()` work ?

File: mappings/net/minecraft/item/CrossbowItem.mapping
Comment (Gegy) [reply mappings/net/minecraft/item/CrossbowItem.mapping]: With `1.0` would it not add velocity up to `0.008` on each axis?

Comment (ChloeDawn) [reply mappings/net/minecraft/item/CrossbowItem.mapping]: Oh.. Yeah, right... I'm bad at math. So, `inaccuracy` then?

Comment (Gegy) [reply mappings/net/minecraft/item/CrossbowItem.mapping]: Yeah. Could also be like `inaccuracyMultiplier` or `inaccuracyFactor`.

File: mappings/net/minecraft/block/dispenser/ItemDispenserBehavior.mapping
Comment (ChloeDawn) [reply mappings/net/minecraft/block/dispenser/ItemDispenserBehavior.mapping:2]: Changed to `dispenseSilently` in 176d9bc, marking as resolved

File: mappings/net/minecraft/item/Item.mapping
Comment (ChloeDawn) [reply mappings/net/minecraft/item/Item.mapping]: Would `inventoryTick` be explicit enough?

Comment (Pyrofab) [reply mappings/net/minecraft/item/Item.mapping]: I think so

File: mappings/net/minecraft/item/ItemStack.mapping
Comment (ChloeDawn) [reply mappings/net/minecraft/item/ItemStack.mapping:10]: The field is named relative to the verbosity of the encapsulation, which is descriptive enough within context imo. The methods would be called as such:
```java
itemStack.isInFrame();
itemStack.getFrame();
itemStack.setFrame(...);
```
Essentially, it is the item's frame (`ItemFrameEntity`).

File: mappings/net/minecraft/item/DefaultFoodSettings.mapping
Comment (ChloeDawn) [reply mappings/net/minecraft/item/DefaultFoodSettings.mapping]: I like `FoodComponent`, it better describes its use in Mojang's step towards a component system for behavioural attributes.

Comment (ChloeDawn) [reply mappings/net/minecraft/item/DefaultFoodSettings.mapping]: On the other hand, they have a hardcoded getter in Item... Not sure if `item.getFoodComponent()` is very nice, and `item.getFood()` would be too vague.

Comment (Pyrofab) [reply mappings/net/minecraft/item/DefaultFoodSettings.mapping]: That's what I was thinking. The tweet introducing this change did mention ECS.

File: mappings/net/minecraft/item/Item.mapping
Comment (ChloeDawn) [reply mappings/net/minecraft/item/Item.mapping]: Maybe `hit`? It handles stack damaging for the implementations in `MiningToolItem`, `HoeItem`, `SwordItem` and `TridentItem`, then returns true. The default return value in `Item` is `false`.

File: mappings/net/minecraft/item/DefaultFoodSettings.mapping
Comment (Pyrofab) [reply mappings/net/minecraft/item/DefaultFoodSettings.mapping]: I don't have a problem with `item.getFoodComponent()` but there may more fluent names

File: mappings/net/minecraft/item/Item.mapping
Comment (ChloeDawn) [reply mappings/net/minecraft/item/Item.mapping]: Resolved by a57ec18

File: mappings/net/minecraft/item/ItemStack.mapping
Comment (ChloeDawn) [reply mappings/net/minecraft/item/ItemStack.mapping:28]: It updates a cached state used internally to avoid redundant checks. If anything it should be `updateCachedEmptyState` but it feels a little verbose for an internal method.

File: mappings/net/minecraft/item/SpawnEggItem.mapping
Comment (ChloeDawn) [reply mappings/net/minecraft/item/SpawnEggItem.mapping]: Resolved by cc5abf1

File: mappings/net/minecraft/item/ItemStack.mapping
Comment (Pyrofab) [reply mappings/net/minecraft/item/ItemStack.mapping:28]: Fair enough

File: mappings/net/minecraft/item/ItemStack.mapping
Comment (Pyrofab) [reply mappings/net/minecraft/item/ItemStack.mapping:10]: Dunno, I still think it would not hurt to be a little more verbose here.

File: mappings/net/minecraft/item/Item.mapping
Comment (Pyrofab) [reply mappings/net/minecraft/item/Item.mapping]: `hit` sounds good to me.

Comment (ChloeDawn) [reply mappings/net/minecraft/item/Item.mapping]: Resolved by bf7a33a, after a discussion on Discord

File: mappings/net/minecraft/item/ItemStack.mapping
Comment (ChloeDawn) [reply mappings/net/minecraft/item/ItemStack.mapping:10]: Resolved until further notice through discussion on Discord
> InsomniaKitten: how else would an item's frame be interpreted

> Pyrofab: Fair

File: mappings/net/minecraft/item/HorseArmorItem.mapping
Comment (ChloeDawn) [reply mappings/net/minecraft/item/HorseArmorItem.mapping:8]: Resolved through brief discussion in Discord. It represents the entity texture of the horse armour and is reasonably descriptive in context. The real concern here is why the hell Mojang store their entity textures in an item class, set up via the constructor... Dammit, Mojang.

File: mappings/net/minecraft/item/DefaultFoodSettings.mapping
Comment (ChloeDawn) [reply mappings/net/minecraft/item/DefaultFoodSettings.mapping]: Resolved by 794bc6e. Feedback is welcome on the getter name in `Item`.

File: mappings/net/minecraft/item/ComplexItem.mapping
Comment (ChloeDawn) [reply mappings/net/minecraft/item/ComplexItem.mapping]: Resolved by d06eb13 after discussion in Discord

File: mappings/net/minecraft/item/CrossbowItem.mapping
Comment (ChloeDawn) [reply mappings/net/minecraft/item/CrossbowItem.mapping]: Resolved by b4a92fd after discussion in Discord

**Comment by modmuss50** [comment]
Just going to pick out some names that I think might need some more feedback on from others. Im currently unsure on them. These are very commonly used so I think getting more feedback is needed.

# ItemStack

- amount -> count

- addAmount -> increment 

- subtractAmount -> decrement 

`count` is the mojang names, but that is by no means a solid reason to use it.

File: mappings/net/minecraft/item/FoodComponent.mapping
Comment (modmuss50) [inline mappings/net/minecraft/item/FoodComponent.mapping:1]: This suffix conflicts with the recent chat changes. I think this is ok as the chat stuff needs more work.

**Comment by ChloeDawn** [comment]
`increment` and `decrement` are not Mojang names, they were names discussed on Discord. What else would you be incrementing other than the count? It doesn't really need to be explicit.

**Comment by modmuss50** [comment]
ah ok, I like `increment` and `decrement`, but im still unsure on count.

It could just be me, thats why im trying to get more feedback on them.

File: mappings/net/minecraft/item/Item.mapping
Comment (modmuss50) [inline mappings/net/minecraft/item/Item.mapping:6]: Is just `group` a bit vague?

Comment (ChloeDawn) [reply mappings/net/minecraft/item/Item.mapping:6]: It's the group (`ItemGroup`) bound to the item. The getter is also `getGroup()`, so you would call as `item.getGroup()` returning an `ItemGroup`.

Comment (modmuss50) [reply mappings/net/minecraft/item/Item.mapping:6]: Ah yeah, thats a fair point.

**Comment by grondag** [comment]
> ah ok, I like `increment` and `decrement`, but im still unsure on count.
> 
> It could just be me, thats why im trying to get more feedback on them.

I like count (and the other two.)

File: mappings/net/minecraft/item/ItemPropertyGetter.mapping
Comment (liach) [inline mappings/net/minecraft/item/ItemPropertyGetter.mapping]: Did enigma glitch here?

File: mappings/net/minecraft/item/ItemPlacementContext.mapping
Comment (liach) [inline mappings/net/minecraft/item/ItemPlacementContext.mapping:2]: `existingReplaceable`? Consider a name that doesn't clash with the method name if possible.

**Review by liach — APPROVED** [review]
lgtm. Since you have named `FoodComponent`, we should consider moving the existing `Component` classes to `Text` instead.

File: mappings/net/minecraft/item/ItemPropertyGetter.mapping
Comment (ChloeDawn) [reply mappings/net/minecraft/item/ItemPropertyGetter.mapping]: Seems so, yes. This is a deobfuscated method and it has somehow generated a mapping for it...

Comment (ChloeDawn) [reply mappings/net/minecraft/item/ItemPropertyGetter.mapping]: ~~Is it safe to just edit this line in the mapping file to fix this @modmuss50?~~

File: mappings/net/minecraft/item/ItemPlacementContext.mapping
Comment (ChloeDawn) [reply mappings/net/minecraft/item/ItemPlacementContext.mapping:2]: Why does it need to differ from the method? The field is private.

Comment (liach) [reply mappings/net/minecraft/item/ItemPlacementContext.mapping:2]: It is fine. Just in case there is a setter, but there is not.

File: mappings/net/minecraft/item/ItemPropertyGetter.mapping
Comment (ChloeDawn) [reply mappings/net/minecraft/item/ItemPropertyGetter.mapping]: Resolved by 9efcc43

**Review by Pyrofab — APPROVED** [review]

**Review by modmuss50 — APPROVED** [review]
Im happy with this, its got some changes that will effect a lot of people, but the mappings migration tool exists.

Nice work.

**Comment by modmuss50** [comment]
Im going to wait for another maintainer to approve this as well, but it seems good.

**Comment by Sollace** [comment]
Based on Discord discussions, I'm going to suggest: `ProjectileShootingItem` -> `RangedWeaponItem`

> **Ranged weapon**
> A ranged weapon is any weapon that can engage targets beyond hand-to-hand distance, i.e. at distances greater than the physical reach of the weapon itself. It is sometimes also called projectile weapon or missile weapon because it typically works by launching projectiles, though technically a directed-energy weapon is also a ranged weapon. In contrast, a weapon intended to be used in hand-to-hand combat is called a melee weapon. -- Wikipedia

**Comment by ChloeDawn** [comment]
> Based on Discord discussions, I'm going to suggest: `ProjectileShootingItem` -> `RangedWeaponItem`
> 
> > **Ranged weapon**
> > A ranged weapon is any weapon that can engage targets beyond hand-to-hand distance, i.e. at distances greater than the physical reach of the weapon itself. It is sometimes also called projectile weapon or missile weapon because it typically works by launching projectiles, though technically a directed-energy weapon is also a ranged weapon. In contrast, a weapon intended to be used in hand-to-hand combat is called a melee weapon. -- Wikipedia

Resolved by ed069d3

**Review by liach — APPROVED** [review]

**Pull request merged**
